### PR TITLE
Fixing prefetch bugs

### PR DESCRIPTION
**Pull request by psuriana**
Add simplifier to remove Prefetch IR when the predicate is always false. Also remove the Prefetch IR, if during `inject_prefetch`, the region being prefetched is empty. 

During `schedule_functions`, placeholder prefetches need to be peeled off; otherwise, the prefetch will end fail to account of function calls by the producer computed within that loop level.

### CODE REVIEW

**Comment by steven-johnson** [comment]
Does any of the code for the Prefetch handler in Simplifier need to get migrated into https://github.com/halide/Halide/pull/2845 as well?

File: src/Prefetch.cpp [hunk 1]
```diff
@@ -148,11 +148,18 @@ class InjectPrefetch : public IRMutator2 {
...
             }
+            internal_assert(!new_bounds.empty());
             return Prefetch::make(op->name, op->types, new_bounds, op->prefetch, condition, std::move(body));
         }
 
         if (!body.same_as(op->body)) {
             return Prefetch::make(op->name, op->types, op->bounds, op->prefetch, op->condition, std::move(body));
+        } else if (op->bounds.empty()) {
+            // Remove the Prefetch IR since it is prefetching an empty region
+            std::cerr << "Warning: Removing prefetch of " << p.name
```
Comment (steven-johnson) [inline src/Prefetch.cpp]: This should use `user_warning` instead of `std::cerr`

**Comment by psuriana** [comment]
Yeah, the Prefetch handler might need to be moved into #2845. It is relatively simple changes though.

**Comment by steven-johnson** [comment]
Breakage is unrelated, landing

**Comment by abadams** [comment]
Breakage doesn't look unrelated to me - it's the prefetch test that's failing for hexagon.

**Pull request merged**